### PR TITLE
Persistent Subscriptions to $all

### PR DESCRIPTION
**Pull request by George-Payne**
## Client
- Add server features support
- Align persistentSubscriptionSettings [breaking change] [RFC022](https://github.com/EventStore/architecture-and-planning/blob/main/rfcs/022%20-%20gRPC%20Persistent%20Subscriptions.md#defaults-and-parameter-descriptions)
- `createPersistentSubscriptionToAll`
- `updatePersistentSubscriptionToAll`
- `deletePersistentSubscriptionToAll`
- `connectToPersistentSubscriptionToAll`
- make `TwoWaySubscription` generic, and rename to` PersistentSubscriptionImpl`

## tests
- Add version check environment for skipping tests based on version
- Update tests to use new `persistentSubscriptionSettings` shape
- Tests for added methods
- Make sure a leader has been elected before starting tests

### CODE REVIEW

File: src/__test__/persistentSubscription/connectToPersistentSubscriptionToAll.test.ts [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+        );
+      }
+    });
+
+    test("Can skip the version check, if forced", async () => {
+      const GROUP_NAME = "oh_boy";
+
+      try {
+        const ps$all = client.connectToPersistentSubscriptionToAll(GROUP_NAME, {
+          skipSupportedCheck: true,
```
Comment (oskardudycz) [inline src/__test__/persistentSubscription/connectToPersistentSubscriptionToAll.test.ts:81]: **QUESTION:** Out of curiosity, when would we want to skip the supported check?

File: src/__test__/persistentSubscription/deletePersistentSubscriptionToAll.test.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      { rootCertificate: node.rootCertificate },
+      { username: "admin", password: "changeit" }
+    );
+  });
+
+  afterAll(async () => {
+    await node.down();
+  });
+
+  optionalDescribe(!supported)("Not Supported (<21.10)", () => {
```
Comment (oskardudycz) [inline src/__test__/persistentSubscription/deletePersistentSubscriptionToAll.test.ts:32]: **CONCERN:** Don't we have persistent subscriptions to `$all` supported from 21.6? See the release notes: https://www.eventstore.com/blog/21.6.0-release-notes. Do we have a reason to limit it to above 21.10?

File: src/persistentSubscription/utils/persistentSubscriptionSettings.ts [hunk 3]
```diff
@@ -84,21 +77,51 @@ export interface PersistentSubscriptionSettings {
...
-  minCheckpointCount: 10,
-  maxCheckpointCount: 1_000,
-  maxSubscriberCount: UNLIMITED,
+  checkPointLowerBound: 10,
+  checkPointUpperBound: 1_000,
+  maxSubscriberCount: UNBOUNDED,
   liveBufferSize: 500,
   readBatchSize: 20,
   historyBufferSize: 500,
   strategy: ROUND_ROBIN,
```
Comment (oskardudycz) [inline src/persistentSubscription/utils/persistentSubscriptionSettings.ts:110]: **SUGGESTION:** As we're aligning terminology, then we could also change it to `consumerStrategyName`.
```suggestion
  consumerStrategyName: ROUND_ROBIN,
```

File: src/__test__/persistentSubscription/deletePersistentSubscriptionToAll.test.ts (hunk 2 above)
Comment (George-Payne) [reply src/__test__/persistentSubscription/deletePersistentSubscriptionToAll.test.ts:32]: As discussed, 21.6 doesn't have support for the server features api, so persistent subscriptions to all with 21.6 are not supported by this client. This dissension will be communicated in release notes.

File: src/__test__/persistentSubscription/connectToPersistentSubscriptionToAll.test.ts [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+  persistentSubscriptionToAllSettingsFromDefaults,
+  START,
+  UnsupportedError,
+  streamNameFilter,
+  END,
+} from "../..";
+
+const asyncPipeline = promisify(pipeline);
+
+describe("connectToPersistentSubscriptionToAll", () => {
```
Comment (oskardudycz) [inline src/__test__/persistentSubscription/connectToPersistentSubscriptionToAll.test.ts:31]: **SUGGESTION:** That might be out of the scope of this PR, but it'd be good to add more tests around handling node loss, reconnecting etc. It's not a blocking suggestion, but as there are always many support issues for that feature, it might be worth enhancing those edge cases coverage.

**Review by oskardudycz — APPROVED** [review]
Nice job @George-Payne 👍

File: src/__test__/persistentSubscription/connectToPersistentSubscriptionToAll.test.ts (hunk 4 above)
Comment (George-Payne) [reply src/__test__/persistentSubscription/connectToPersistentSubscriptionToAll.test.ts:31]: Agreed, I've opened an issue here: #231

File: src/__test__/utils/Cluster.ts [hunk 5]
```diff
@@ -327,6 +332,46 @@ export class Cluster {
...
+          const response = await exec(
+            node,
+            `curl --fail --insecure http${
+              this.insecure ? "" : "s"
+            }://localhost:2113/gossip`,
+            { cwd: this.path() }
+          );
+
+          interface Member {
+            state: "Unknown" | "Leader" | "follower";
```
Comment (pvanbuijtene) [inline src/__test__/utils/Cluster.ts:356]: Typo?
```suggestion
            state: "Unknown" | "Leader" | "Follower";
```

File: src/persistentSubscription/utils/persistentSubscriptionSettings.ts [hunk 6]
```diff
@@ -81,24 +74,54 @@ export interface PersistentSubscriptionSettings {
...
+}
+
+const defaults: PersistentSubscriptionSettingsGeneric = {
   resolveLinkTos: false,
-  extraStats: false,
-  fromRevision: START,
+  extraStatistics: false,
   messageTimeout: 30_000,
   maxRetryCount: 10,
   checkpointAfter: 2_000,
```
Comment (pvanbuijtene) [inline src/persistentSubscription/utils/persistentSubscriptionSettings.ts:103]: nitpick, not sure how strict we want to be 😬 
```suggestion
  checkPointAfter: 2_000,
```

**Review by pvanbuijtene — APPROVED** [review]
